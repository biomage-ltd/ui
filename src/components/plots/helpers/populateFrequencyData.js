const populateFrequencyData = (spec, hierarchy, properties, config) => {
  const calculateSum = (proportionClusters, xAxisClustersIds) => {
    let sum = 0;

    if (!xAxisClustersIds.length) {
      proportionClusters.forEach((cellSetCluster) => {
        sum += properties[cellSetCluster.key].cellIds.size;
      });
      return sum;
    }
    proportionClusters.forEach((cellSetCluster) => {
      const cellSetIds = Array.from(properties[cellSetCluster.key].cellIds);
      sum += xAxisClustersIds.filter((id) => cellSetIds.includes(id)).length;
    });
    return sum;
  };

  const getClustersForGrouping = (name) => hierarchy.filter((cluster) => (
    cluster.key === name))[0]?.children;

  const populateData = (x, y, cluster, sum, data) => {
    let value = y;
    if (config.frequencyType === 'proportional') {
      value = (y / sum) * 100;
    }

    data.push({
      x,
      y: value,
      c: cluster.name,
      color: cluster.color,
    });

    return data;
  };

  let data = [];

  const proportionClusters = hierarchy.filter((cluster) => (
    cluster.key === config.proportionGrouping))[0]?.children;

  if (!proportionClusters) {
    return [];
  }

  const clustersInXAxis = getClustersForGrouping(config.xAxisGrouping);
  if (!clustersInXAxis) {
    const sum = calculateSum(proportionClusters, []);
    proportionClusters.forEach((clusterName) => {
      const x = 1;
      const y = properties[clusterName.key].cellIds.size;
      const cluster = properties[clusterName.key];
      data = populateData(x, y, cluster, sum, data);
    });
  } else {
    clustersInXAxis.forEach((clusterInXAxis) => {
      const clusterInXAxisIds = Array.from(properties[clusterInXAxis.key].cellIds);

      const sum = calculateSum(proportionClusters, clusterInXAxisIds);

      proportionClusters.forEach((clusterName) => {
        const x = properties[clusterInXAxis.key].name;
        const cellSetIds = Array.from(properties[clusterName.key].cellIds);
        const y = clusterInXAxisIds.filter((id) => cellSetIds.includes(id)).length;
        const cluster = properties[clusterName.key];

        if (y !== 0) {
          data = populateData(x, y, cluster, sum, data);
        }
      });
    });
  }

  spec.data.forEach((datum) => {
    datum.values = data; // eslint-disable-line no-param-reassign
  });
  return data;
};

export default populateFrequencyData;
