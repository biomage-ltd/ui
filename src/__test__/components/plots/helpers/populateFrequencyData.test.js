import populateFrequencyData from '../../../../components/plots/helpers/populateFrequencyData';

describe('populateFrequencyData', () => {
  let configMock = {};
  let specMock = {};
  let hierarchyMock = [];
  let propertiesMock = {};

  beforeEach(() => {
    specMock = {
      data: [
        {},
      ],
    };

    hierarchyMock = [
      { key: 'sample', children: [{ key: 'sample1' }, { key: 'sample2' }, { key: 'sample3' }] },
      { key: 'louvain', children: [{ key: 'louvain1' }, { key: 'louvain2' }] },
    ];

    propertiesMock = {
      sample1: {
        cellIds: new Set([1, 2, 3, 4]),
        color: 'sample1',
        name: '1',
      },
      sample2: {
        cellIds: new Set([5, 6, 7]),
        color: 'sample2',
        name: '2',
      },
      sample3: {
        cellIds: new Set([8, 9, 10]),
        color: 'sample3',
        name: '3',
      },
      louvain1: {
        cellIds: new Set([1, 2, 5, 8, 10]),
        color: 'louvain1',
        name: '1',
      },
      louvain2: {
        cellIds: new Set([3, 4, 6, 7, 9]),
        color: 'louvain2',
        name: '2',
      },
    };
  });

  test('with louvain as xAxis and sample as proportion', () => {
    configMock = { proportionGrouping: 'sample', xAxisGrouping: 'louvain', frequencyType: 'proportional' };

    populateFrequencyData(specMock, hierarchyMock, propertiesMock, configMock);

    // Should populate with 2 bars, one for each of the louvain clusters
    expect(specMock.data).toMatchSnapshot();
  });

  test('with sample as xAxis and louvain as proportion', () => {
    configMock = { proportionGrouping: 'louvain', xAxisGrouping: 'sample', frequencyType: 'proportional' };

    populateFrequencyData(specMock, hierarchyMock, propertiesMock, configMock);

    // Should populate with 3 bars, one for each of the samples
    expect(specMock.data).toMatchSnapshot();
  });

  test('with empty xAxis and sample as proportion we only get data for sample', () => {
    configMock = { proportionGrouping: 'sample', xAxisGrouping: 'louvain', frequencyType: 'proportional' };

    // Remove grouping that is in xAxis
    hierarchyMock = hierarchyMock.filter((rootNode) => rootNode.key !== 'louvain');

    populateFrequencyData(specMock, hierarchyMock, propertiesMock, configMock);

    // Should populate only with the 3 bars for sample
    expect(specMock.data).toMatchSnapshot();
  });

  test('with empty proportion', () => {
    configMock = { proportionGrouping: 'sample', xAxisGrouping: 'louvain', frequencyType: 'proportional' };

    // Remove grouping that is in proportion
    hierarchyMock = hierarchyMock.filter((rootNode) => rootNode.key !== 'sample');

    populateFrequencyData(specMock, hierarchyMock, propertiesMock, configMock);

    // Should not populate any data
    expect(specMock.data).toMatchSnapshot();
  });

  it('displays data with count instead of percentages when the frequencyType is count', () => {
    configMock = { proportionGrouping: 'sample', xAxisGrouping: 'louvain', frequencyType: 'count' };

    populateFrequencyData(specMock, hierarchyMock, propertiesMock, configMock);

    // Should populate with 2 bars, one for each of the louvain clusters with count
    expect(specMock.data).toMatchSnapshot();
  });
});
