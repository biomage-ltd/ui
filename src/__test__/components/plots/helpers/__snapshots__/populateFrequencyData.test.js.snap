// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populateFrequencyData displays data with count instead of percentages when the frequencyType is count 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "c": "1",
        "color": "sample1",
        "x": "1",
        "y": 2,
      },
      Object {
        "c": "2",
        "color": "sample2",
        "x": "1",
        "y": 1,
      },
      Object {
        "c": "3",
        "color": "sample3",
        "x": "1",
        "y": 2,
      },
      Object {
        "c": "1",
        "color": "sample1",
        "x": "2",
        "y": 2,
      },
      Object {
        "c": "2",
        "color": "sample2",
        "x": "2",
        "y": 2,
      },
      Object {
        "c": "3",
        "color": "sample3",
        "x": "2",
        "y": 1,
      },
    ],
  },
]
`;

exports[`populateFrequencyData with empty proportion 1`] = `
Array [
  Object {},
]
`;

exports[`populateFrequencyData with empty xAxis and sample as proportion we only get data for sample 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "c": "1",
        "color": "sample1",
        "x": 1,
        "y": 40,
      },
      Object {
        "c": "2",
        "color": "sample2",
        "x": 1,
        "y": 30,
      },
      Object {
        "c": "3",
        "color": "sample3",
        "x": 1,
        "y": 30,
      },
    ],
  },
]
`;

exports[`populateFrequencyData with louvain as xAxis and sample as proportion 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "c": "1",
        "color": "sample1",
        "x": "1",
        "y": 40,
      },
      Object {
        "c": "2",
        "color": "sample2",
        "x": "1",
        "y": 20,
      },
      Object {
        "c": "3",
        "color": "sample3",
        "x": "1",
        "y": 40,
      },
      Object {
        "c": "1",
        "color": "sample1",
        "x": "2",
        "y": 40,
      },
      Object {
        "c": "2",
        "color": "sample2",
        "x": "2",
        "y": 40,
      },
      Object {
        "c": "3",
        "color": "sample3",
        "x": "2",
        "y": 20,
      },
    ],
  },
]
`;

exports[`populateFrequencyData with sample as xAxis and louvain as proportion 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "c": "1",
        "color": "louvain1",
        "x": "1",
        "y": 50,
      },
      Object {
        "c": "2",
        "color": "louvain2",
        "x": "1",
        "y": 50,
      },
      Object {
        "c": "1",
        "color": "louvain1",
        "x": "2",
        "y": 33.33333333333333,
      },
      Object {
        "c": "2",
        "color": "louvain2",
        "x": "2",
        "y": 66.66666666666666,
      },
      Object {
        "c": "1",
        "color": "louvain1",
        "x": "3",
        "y": 66.66666666666666,
      },
      Object {
        "c": "2",
        "color": "louvain2",
        "x": "3",
        "y": 33.33333333333333,
      },
    ],
  },
]
`;
